dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/predictor/PredictionPreprocessor.java

```diff
@@ -0,0 +1,205 @@
+
+        return deg.getMembers().stream()
+            .map( de -> clonePredictor( predictor, preEx, de ) )
+            .collect( toList() );
+    }
+
+    /**
+     * Clones a predictor for a particular data element.
+     * <p>
+     * Note: The .toBuilder().build() pattern makes a shallow clone. In the
+     * first attempt at writing this method, SerializationUtils.clone() was used
+     * to make a deep clone, but that didn't work because some nested objects
+     * were Hibernate proxies and the resulting objects triggered unable to
+     * initialize proxy exceptions.
+     */
+    private Predictor clonePredictor( Predictor predictor, PredictorPreprocessorExpression preEx,
+        DataElement dataElement )
+    {
+        Predictor clone = predictor.toBuilder().build();
+
+        clone.setOutput( dataElement );
+
+        String mainExpression = preEx.getMain();
+        String variable = preEx.getVariable();
+        String clonedExpression = mainExpression.replace( variable, dataElement.getUid() );
+        Expression clonedGenerator = new Expression( clonedExpression, clone.getGenerator().getDescription(),
+            clone.getGenerator().getMissingValueStrategy() );
+        clone.setGenerator( clonedGenerator );
+
+        return clone;
+    }
+
+    /**
+     * Gets the description for a predictor.
+     */
+    private String describe( String expression )
+    {
+        PredictorPreprocessorExpression preEx = new PredictorPreprocessorExpression( expression );
+
+        // Get prefix description (resolve data element group)
+        DataElementGroup deg = getDeg( preEx.getDegUid(), expression );
+        String prefixDescription = preEx.getPrefix().replace( preEx.getTaggedDegUid(), deg.getName() );
+
+        // Find item descriptions for the main expression with a dummy
+        // expression resolving the data element variable to some DE in the DEG
+        DataElement de = getOneDataElement( deg, expression );
+        String mainExpressionWithOneDataElement = preEx.getMain().replace( preEx.getVariable(), de.getUid() );
+        Map<String, String> itemDescriptions = expressionService.getExpressionItemDescriptions(
+            mainExpressionWithOneDataElement, PREDICTOR_EXPRESSION );
+
+        // Get the main expression description
+        String mainDescription = preEx.getMain();
+        for ( Map.Entry<String, String> entry : itemDescriptions.entrySet() )
+        {
+            mainDescription = mainDescription.replace( entry.getKey(), entry.getValue() );
+        }
+
+        // Return the complete description
+        return prefixDescription + PREPROCESSOR_SEPARATOR + mainDescription;
+    }
+
+    /**
+     * Gets the data element group for preprocessing.
+     */
+    private DataElementGroup getDeg( String deUid, String expression )
+    {
+        DataElementGroup deg = idObjectManager.get( DataElementGroup.class, deUid );
+
+        if ( deg == null )
+        {
+            throw new ParserException(
```

Comment:
Any ID lookup should result in a `org.hisp.dhis.feedback.NotFoundException` if we don't want to catch it so it ends up as the response to the client.

Reply:
I tried that, and it got real messy with declaring the `throws` on each of the nested calls. All the other expression description processing generates runtime errors only and so this needed special handling. As a compromise I changed this to `IllegalStateException` for now which is also used elsewhere in expression description generation and is trapped well in the methods returning expression status.

Speaking of which, currently in the controllers for expression descriptions we first do the check to see if an expression is valid which in turn calls the expression description generation. Then if it is valid, we call the expression description generation again to get the expression. Hopefully when we integrate the PEG parser we can make this much less convoluted and just generate the expression or the exception in one go and not have to do things twice.